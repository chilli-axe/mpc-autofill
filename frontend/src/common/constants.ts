import { CardType as CardTypeSchema } from "@/common/schema_types";
import { Cardstock, CardType, Faces, SortBy } from "@/common/types";
export const ProjectName = "MPC Autofill";
export const MakePlayingCards = "MakePlayingCards.com";
export const MakePlayingCardsURL = "https://www.makeplayingcards.com";

export const Card: CardType = CardTypeSchema.Card;
export const Cardback: CardType = CardTypeSchema.Cardback;
export const Token: CardType = CardTypeSchema.Token;

export const SelectedImageSeparator = "@";
export const CardTypeSeparator = ":";
export const FaceSeparator = "//";

export const CardTypePrefixes: { [prefix: string]: CardType } = {
  "": Card,
  b: Cardback,
  t: Token,
};

export const ReversedCardTypePrefixes = Object.fromEntries(
  Object.keys(CardTypePrefixes).map((prefix: string) => [
    CardTypePrefixes[prefix],
    prefix.length > 0 ? prefix + CardTypeSeparator : prefix,
  ])
);

export const Front: Faces = "front";
export const Back: Faces = "back";

export const ToggleButtonHeight = 38; // pixels
export const NavbarHeight = 50; // pixels - aligns with the natural height of the navbar
export const RibbonHeight = 54; // pixels
export const NavbarLogoHeight = 40; // pixels
export const ContentMaxWidth = 1200; // pixels - aligns with bootstrap's large breakpoint

export const MinimumDPI = 0;
export const MaximumDPI = 1500;
export const DPIStep = 50;

export const MaximumSize = 30; // megabytes
export const SizeStep = 1;

export const CSRFKey = "csrftoken";
export const SearchSettingsKey = "searchSettings";
export const GoogleAnalyticsConsentKey = "googleAnalyticsConsent";
export const BackendURLKey = "backendURL";

export const Brackets: Array<number> = [
  18, 36, 55, 72, 90, 108, 126, 144, 162, 180, 198, 216, 234, 396, 504, 612,
];

export const ProjectMaxSize: number = Brackets[Brackets.length - 1];

export enum QueryTags {
  BackendSpecific = "backendSpecific",
  SearchResults = "searchResults",
  SampleCards = "sampleCards",
}

export const S27: Cardstock = "(S27) Smooth";
export const S30: Cardstock = "(S30) Standard Smooth";
export const S33: Cardstock = "(S33) Superior Smooth";
export const M31: Cardstock = "(M31) Linen";
export const P10: Cardstock = "(P10) Plastic";
export const Cardstocks: Array<Cardstock> = [S27, S30, S33, M31, P10];

export const CardstockFoilCompatibility: { [cardstock in Cardstock]: boolean } =
  {
    [S27]: true,
    [S30]: true,
    [S33]: true,
    [M31]: true,
    [P10]: false,
  };

export const GoogleDriveImageAPIURL =
  "https://script.google.com/macros/s/AKfycbw8laScKBfxda2Wb0g63gkYDBdy8NWNxINoC4xDOwnCQ3JMFdruam1MdmNmN4wI5k4/exec";

export const SearchResultsEndpointPageSize = 300;
export const CardEndpointPageSize = 1000;

export enum CSVHeaders {
  quantity = "Quantity",
  frontQuery = "Front",
  frontSelectedImage = "Front ID",
  backQuery = "Back",
  backSelectedImage = "Back ID",
}

export const ExploreDebounceMS = 700;
export const ExplorePageSize = 60;

export const SortByOptions: { [option in SortBy]: string } = {
  dateCreatedDescending: "Date Created (Newest-Oldest)",
  dateCreatedAscending: "Date Created (Oldest-Newest)",
  dateModifiedDescending: "Date Modified (Newest-Oldest)",
  dateModifiedAscending: "Date Modified (Oldest-Newest)",
  nameAscending: "Name (A-Z)",
  nameDescending: "Name (Z-A)",
};
