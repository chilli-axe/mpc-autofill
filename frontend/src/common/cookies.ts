/**
 * Data API for interacting with anything stored in cookies or local storage.
 */

// import Ajv2020 from "ajv/dist/2020";
import Cookies from "js-cookie";

import {
  BackendURLKey,
  CSRFKey,
  GoogleAnalyticsConsentKey,
  SearchSettingsKey,
} from "@/common/constants";
import { Convert } from "@/common/schema_types";
import { SearchSettings, SourceDocuments, SourceRow } from "@/common/types";
import { getSourceRowsFromSourceSettings } from "@/common/utils";
import { getDefaultSearchSettings } from "@/store/slices/searchSettingsSlice";

// import * as SearchSettingsSchema from "../../../schemas/schemas/SearchSettings.json";

// const ajv = new Ajv2020();

//# region CSRF
// TODO: unsure if we still need this.

export function getCSRFHeader(): HeadersInit | undefined {
  const csrfToken = Cookies.get(CSRFKey);
  if (csrfToken != null) {
    return { "X-CSRFToken": csrfToken };
  }
  return undefined;
}

//# endregion

//# region search settings

/**
 * Get search settings from localStorage data. If valid data is retrieved,
 * ensure that all `sourceDocuments` are included in the returned settings,
 * with any new sources that weren't previously included added to the end and enabled.
 */
export function getLocalStorageSearchSettings(
  sourceDocuments: SourceDocuments
): SearchSettings {
  const serialisedRawSettings = localStorage.getItem(SearchSettingsKey) ?? "{}";
  try {
    const searchSettings = Convert.toSearchSettings(serialisedRawSettings);
    // great, the user has valid search settings stored in their browser local storage.
    // reconcile against sourceDocuments
    const sourceInDatabaseSet: Set<number> = new Set(
      Object.values(sourceDocuments).map((sourceDocument) =>
        parseInt(sourceDocument.pk)
      )
    );
    // types have to be narrowed here because quicktype doesn't support our SourceRow data structure :(
    const sources: Array<SourceRow> = getSourceRowsFromSourceSettings(
      searchSettings.sourceSettings
    );
    const sourceInLocalStorageSet: Set<number> = new Set(
      sources.map((row) => row[0])
    );
    // one fat line of reconciliation, good luck reading this future nick! i wrote this at 12:26am.
    searchSettings.sourceSettings.sources = sources
      .filter((row: SourceRow) => sourceInDatabaseSet.has(row[0]))
      .concat(
        Array.from(sourceInDatabaseSet)
          .filter((pk: number) => !sourceInLocalStorageSet.has(pk))
          .map((pk: number) => [pk, true])
      );
    return searchSettings;
  } catch (e) {
    // quicktype will throw an error if the user's stored settings do not match the schema
    // e.g. upon first page load
    // just return the default settings
    return getDefaultSearchSettings(sourceDocuments);
  }
}

export function setLocalStorageSearchSettings(settings: SearchSettings): void {
  localStorage.setItem(SearchSettingsKey, JSON.stringify(settings));
}

//# endregion

//# region google analytics consent

export function getGoogleAnalyticsConsent(): boolean | undefined {
  const rawConsent = Cookies.get(GoogleAnalyticsConsentKey);
  return rawConsent != undefined ? JSON.parse(rawConsent) === true : undefined;
}

export function setGoogleAnalyticsConsent(consent: boolean): void {
  Cookies.set(GoogleAnalyticsConsentKey, JSON.stringify(consent), {
    expires: 365,
    sameSite: "strict",
  });
}

//# endregion

//# region backend

export function getLocalStorageBackendURL() {
  return localStorage.getItem(BackendURLKey);
}

export function setLocalStorageBackendURL(url: string) {
  localStorage.setItem(BackendURLKey, url);
}

export function clearLocalStorageBackendURL(): void {
  localStorage.removeItem(BackendURLKey);
}

//# endregion
