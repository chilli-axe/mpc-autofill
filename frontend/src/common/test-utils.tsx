/**
 * This module contains a few utils for testing with Redux and React Testing Library.
 * Retrieved from https://redux.js.org/usage/writing-tests
 */

import { Store } from "@reduxjs/toolkit";
import { within } from "@testing-library/dom";
import type { RenderOptions } from "@testing-library/react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import FileSaver from "file-saver";
import { MemoryRouterProvider } from "next-router-mock/MemoryRouterProvider";
import React, { PropsWithChildren } from "react";
import { Provider } from "react-redux";

import { localBackendURL } from "@/common/test-constants";
import { Faces } from "@/common/types";
import { LayoutWithoutReduxProvider } from "@/features/ui/Layout";
import { setURL } from "@/store/slices/backendSlice";
import { RootState, setupStore } from "@/store/store";

//# region redux test setup

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: Partial<RootState>;
  store?: Store;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  configureLocalBackend: boolean = true
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (store != undefined) {
      if (configureLocalBackend) {
        store.dispatch(setURL(localBackendURL));
      }
      return (
        <MemoryRouterProvider>
          <Provider store={store}>
            <LayoutWithoutReduxProvider>{children}</LayoutWithoutReduxProvider>
          </Provider>
        </MemoryRouterProvider>
      );
    } else {
      return (
        <MemoryRouterProvider>
          <LayoutWithoutReduxProvider>{children}</LayoutWithoutReduxProvider>
        </MemoryRouterProvider>
      );
    }
  }

  // Return an object with the store and all of RTL's query functions
  // @ts-ignore
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

//# endregion

//# region assertions

export async function expectCardSlotToExist(slot: number) {
  // front and back are both picked up by this query
  await waitFor(() =>
    expect(screen.getAllByText(`Slot ${slot}`)).toHaveLength(2)
  );
}

export async function expectCardSlotToNotExist(slot: number) {
  await waitFor(() =>
    expect(screen.queryAllByText(`Slot ${slot}`)).toHaveLength(0)
  );
}

/**
 * This function helps with asserting that a particular card slot exists
 * and has the expected image selected.
 */
async function expectCardSlotState(
  testId: string,
  cardName?: string | null,
  selectedImage?: number | null,
  totalImages?: number | null
) {
  await waitFor(() => expect(screen.getByTestId(testId)).toBeInTheDocument());
  const cardElement = screen.getByTestId(testId);
  if (cardName === null) {
    await waitFor(() =>
      expect(
        within(cardElement).getByText("Your search query")
      ).toBeInTheDocument()
    );
  } else if (cardName !== undefined) {
    await waitFor(() =>
      expect(within(cardElement).getByText(cardName)).toBeInTheDocument()
    );
  }
  if (selectedImage != null && totalImages != null) {
    await waitFor(() =>
      expect(
        within(cardElement).getByText(`${selectedImage} / ${totalImages}`)
      ).toBeInTheDocument()
    );
  }
}

export async function expectCardGridSlotState(
  slot: number,
  face: Faces,
  cardName?: string | null,
  selectedImage?: number | null,
  totalImages?: number | null,
  slotSelected?: boolean | null
) {
  const testId = `${face}-slot${slot - 1}`;
  // note: the specified `slot` should be 1-indexed
  await expectCardSlotState(testId, cardName, selectedImage, totalImages);
  if (slotSelected != null) {
    const cardElement = screen.getByTestId(testId);
    const labelText = `${face}${slot - 1}-${slotSelected ? "" : "un"}checked`;
    await waitFor(() =>
      expect(within(cardElement).getByLabelText(labelText)).toBeInTheDocument()
    );
  }
}

export async function expectCardbackSlotState(
  cardName?: string,
  selectedImage?: number,
  totalImages?: number
) {
  return await expectCardSlotState(
    "common-cardback",
    cardName,
    selectedImage,
    totalImages
  );
}

//# endregion

//# region UI interactions

/**
 * Note: the `ping` function from Ping.js must be mocked in tests where you call this function.
 * We cannot use MSW to intercept Ping.js because that library works by creating an `Image` with
 * `src` of the site to ping's `favicon.ico`, and MSW can't intercept requests to load images.
 */
export async function configureBackend(url: string) {
  await waitFor(() => screen.getByLabelText("configure-server-btn").click());

  const backendOffcanvas = await waitFor(() =>
    screen.getByTestId("backend-offcanvas")
  );

  const textField = await waitFor(() =>
    within(backendOffcanvas).getByLabelText("backend-url")
  );
  const user = userEvent.setup();
  await user.clear(textField);
  if (url !== "") {
    await user.type(textField, url);
  }
  within(backendOffcanvas).getByLabelText("submit-backend-url").click();
  await waitFor(() =>
    expect(
      within(backendOffcanvas).getByText("You're currently connected to", {
        exact: false,
      })
    ).toBeInTheDocument()
  );
  within(backendOffcanvas).getByLabelText("Close").click();
}

export async function configureDefaultBackend() {
  return await configureBackend(localBackendURL);
}

export async function getErrorToast(): Promise<HTMLElement> {
  const errorToast = await waitFor(
    () => screen.getByText("An Error Occurred").parentElement?.parentElement
  );
  if (errorToast == null) {
    throw new Error("error toast is not present");
  }
  return errorToast;
}

export function getAddCardsMenu() {
  return within(screen.getByTestId("right-panel")).getByText("Add Cards", {
    exact: false,
  });
}

export async function openImportTextModal() {
  // open the modal and find the text area
  const addCardsMenu = getAddCardsMenu();
  addCardsMenu.click();
  await waitFor(() => screen.getByText("Text", { exact: false }).click());
  await waitFor(() => expect(screen.getByText("Add Cards — Text")));
  await waitFor(() =>
    expect(screen.getByLabelText("import-text-submit")).not.toBeDisabled()
  ); // here, we wait for DFC pairs to be loaded
  return screen.getByLabelText("import-text");
}

export async function importText(text: string) {
  const textArea = await openImportTextModal();
  const user = userEvent.setup();
  await user.clear(textArea);
  if (text !== "") {
    await user.type(textArea, text);
  }
  screen.getByLabelText("import-text-submit").click();
}

export async function openImportCSVModal() {
  // open the modal and find the upload dropzone
  const addCardsMenu = getAddCardsMenu();
  addCardsMenu.click();
  await waitFor(() => screen.getByText("CSV", { exact: false }).click());
  await waitFor(() => expect(screen.getByText("Add Cards — CSV")));
  const dropzone = screen.getByLabelText("import-csv");
  await waitFor(() => expect(dropzone).not.toBeDisabled()); // here, we wait for DFC pairs to be loaded
  return dropzone.querySelector("input")!;
}

export async function importCSV(fileContents: string) {
  const user = userEvent.setup();
  const dropzone = await openImportCSVModal();
  const file = new File([fileContents], "test.csv", { type: "text/csv" });
  await user.upload(dropzone, [file]);
}

export async function openImportXMLModal() {
  // open the modal and find the upload dropzone
  const addCardsMenu = getAddCardsMenu();
  addCardsMenu.click();
  await waitFor(() => screen.getByText("XML", { exact: false }).click());
  await waitFor(() => expect(screen.getByText("Add Cards — XML")));
  const dropzone = screen.getByLabelText("import-xml");
  await waitFor(() => expect(dropzone).not.toBeDisabled()); // here, we wait for DFC pairs to be loaded
  return dropzone.querySelector("input")!;
}

export async function importXML(
  fileContents: string,
  useXMLCardback: boolean = false
) {
  const user = userEvent.setup();
  const dropzone = await openImportXMLModal();

  const file = new File([fileContents], "test.xml", {
    type: "text/xml;charset=utf-8",
  });
  if (useXMLCardback) {
    await waitFor(() => screen.getByText("Retain Selected Cardback").click());
  }

  await user.upload(dropzone, [file]);
}

export async function openImportURLModal() {
  // open the modal and find the text input
  const addCardsMenu = getAddCardsMenu();
  addCardsMenu.click();
  await waitFor(() => screen.getByText("URL", { exact: false }).click());
  await waitFor(() => expect(screen.getByText("Add Cards — URL")));
  return screen.getByLabelText("import-url");
}

async function openGridSelector(
  cardSlotTestId: string,
  gridSelectorTestId: string,
  selectedImage: number,
  totalImages: number
) {
  expect(totalImages).toBeGreaterThan(1);
  await expectCardSlotState(
    cardSlotTestId,
    undefined,
    selectedImage,
    totalImages
  );

  await waitFor(() =>
    within(screen.getByTestId(cardSlotTestId))
      .getByText(`${selectedImage} / ${totalImages}`)
      .click()
  );
  await waitFor(() => expect(screen.getByText("Option 1")));

  return screen.getByTestId(gridSelectorTestId);
}

export async function openChangeQueryModal(
  cardSlotTestId: string,
  cardName: string
) {
  await waitFor(() =>
    within(screen.getByTestId(cardSlotTestId)).getByText(cardName).click()
  );
  return screen.getByTestId("change-query-modal");
}

export async function changeQuery(
  cardSlotTestId: string,
  cardName: string,
  newQuery: string
) {
  const modal = await openChangeQueryModal(cardSlotTestId, cardName);
  await changeQueries(newQuery);
}

export async function openCardSlotGridSelector(
  slot: number,
  face: Faces,
  selectedImage: number,
  totalImages: number
) {
  return await openGridSelector(
    `${face}-slot${slot - 1}`,
    `${face}-slot${slot - 1}-grid-selector`,
    selectedImage,
    totalImages
  );
}

export async function selectSlot(
  slot: number,
  face: Faces,
  clickType: "double" | "shift" | null = null
) {
  const cardElement = screen.getByTestId(`${face}-slot${slot - 1}`);
  const user = userEvent.setup();
  const element = within(cardElement).getByLabelText(
    `select-${face}${slot - 1}`
  )!.children[0];
  if (clickType === "double") {
    fireEvent.click(element, { detail: 2 });
  } else if (clickType === "shift") {
    await user.keyboard("{Shift>}");
    await user.click(element);
    await user.keyboard("{/Shift}");
  } else {
    await user.click(element);
  }
  await waitFor(() =>
    expect(
      within(cardElement).getByLabelText(`select-${face}${slot - 1}`)!
        .children[0]
    ).toHaveClass("bi-check-square")
  );
}

export async function deselectSlot(slot: number, face: Faces) {
  const cardElement = screen.getByTestId(`${face}-slot${slot - 1}`);
  const user = userEvent.setup();
  await user.click(
    within(cardElement).getByLabelText(`select-${face}${slot - 1}`)!.children[0]
  );
  await waitFor(() =>
    expect(
      within(cardElement).getByLabelText(`select-${face}${slot - 1}`)!
        .children[0]
    ).toHaveClass("bi-square")
  );
}

export function clickMoreSelectOptionsDropdown() {
  screen.getByTestId("more-select-options").click();
}

export async function selectSimilar() {
  clickMoreSelectOptionsDropdown();
  await waitFor(() => screen.getByText("Select Similar").click());
}

export async function selectAll() {
  clickMoreSelectOptionsDropdown();
  await waitFor(() => screen.getByText("Select All").click());
}

export async function changeQueries(query: string) {
  const textField = await waitFor(() =>
    screen.getByLabelText("change-selected-image-queries-text")
  );
  const user = userEvent.setup();
  await user.clear(textField);
  if (query !== "") {
    await user.type(textField, query);
  }
  screen.getByLabelText("change-selected-image-queries-submit").click();
}

export async function changeQueryForSelectedImages(query: string) {
  clickMoreSelectOptionsDropdown();
  await waitFor(() => screen.getByText("Change Query").click());
  await changeQueries(query);
}

export async function changeImageForSelectedImages(cardName: string) {
  clickMoreSelectOptionsDropdown();
  await waitFor(() => screen.getByText("Change Version").click());
  await waitFor(() => expect(screen.getByText("Option 1")));
  await waitFor(() =>
    within(screen.getByTestId("bulk-grid-selector"))
      .getByAltText(cardName)
      .click()
  );
}

export async function clearQueriesForSelectedImages() {
  clickMoreSelectOptionsDropdown();
  await waitFor(() => screen.getByText("Clear Query").click());
}

export async function deleteSelectedImages() {
  clickMoreSelectOptionsDropdown();
  await waitFor(() => screen.getByText("Delete Cards").click());
}

export async function openSearchSettingsModal() {
  screen.getByText(/Search Settings/).click();
  await waitFor(() => expect(screen.getByText("Search Settings")));
  return screen.getByTestId("search-settings");
}

function getDownloadMenu() {
  return within(screen.getByTestId("right-panel")).getByText("Download", {
    exact: false,
  });
}

async function downloadFile(testId: string) {
  // @ts-ignore
  jest.spyOn(global, "Blob").mockImplementation(function (content, options) {
    return { content, options };
  });
  const downloadMenu = getDownloadMenu();
  downloadMenu.click();
  await waitFor(() => screen.getByTestId(testId).click());
  expect(FileSaver.saveAs).toHaveBeenCalledTimes(1);

  // @ts-ignore
  const [blob, filename] = (FileSaver.saveAs as jest.Mock).mock.calls[0];
  return [blob, filename];
}

export async function downloadXML() {
  return await downloadFile("export-xml-button");
}

export async function downloadDecklist() {
  return await downloadFile("export-decklist-button");
}

//# endregion

//# region misc

/**
 * Pretty gross, but useful for asserting that generated XMLs are as expected.
 */
export function normaliseString(text: string): string {
  return text.replaceAll(" ", "").replaceAll("\n", "").replaceAll("\r", "");
}

//# endregion
