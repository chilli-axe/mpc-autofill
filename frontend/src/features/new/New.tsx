import React, { useState } from "react";
import Button from "react-bootstrap/Button";
import Row from "react-bootstrap/Row";
import styled from "styled-components";

import { NewCardsFirstPage } from "@/common/types";
import { Spinner } from "@/components/Spinner";
import { DatedCard } from "@/features/card/Card";
import {
  useGetNewCardsFirstPageQuery,
  useGetNewCardsPageQuery,
} from "@/store/api";
import { useProjectName } from "@/store/slices/backendSlice";

const InlineHeader = styled.h3`
  display: inline;
`;

const InlineParagraph = styled.p`
  display: inline;
`;

function NewCardsForSource({
  sourceKey,
  firstPage,
}: {
  sourceKey: string;
  firstPage: NewCardsFirstPage;
}) {
  // django pagination begins at 1
  const [pageCounter, setPageCounter] = useState<number>(1);
  const getNewCardsPageQuery = useGetNewCardsPageQuery([
    sourceKey,
    pageCounter,
  ]);

  const loadMoreButton = (
    <div className="d-grid gap-0 mx-auto" style={{ maxWidth: 20 + "%" }}>
      <Button
        onClick={() => setPageCounter(pageCounter + 1)}
        disabled={getNewCardsPageQuery.isFetching}
      >
        {getNewCardsPageQuery.isFetching ? <Spinner size={1.5} /> : "Load More"}
      </Button>
    </div>
  );

  return (
    <>
      <InlineHeader className="orpheus">
        <em>{firstPage.source.name}</em>
      </InlineHeader>
      &nbsp;&nbsp;&nbsp;
      <InlineParagraph className="text-primary">
        {firstPage.hits} new card{firstPage.hits != 1 && "s"}
      </InlineParagraph>
      <Row xxl={6} lg={4} md={3} sm={2} xs={2} className="g-0">
        {firstPage.cards.map((card) => (
          <DatedCard
            cardDocument={card}
            headerDate="created"
            key={`whats-new-card-${card.identifier}`}
          />
        ))}
        {(getNewCardsPageQuery.data ?? []).map((card) => (
          <DatedCard
            cardDocument={card}
            headerDate="created"
            key={`whats-new-card-${card.identifier}`}
          />
        ))}
      </Row>
      <br />
      {pageCounter < firstPage.pages && loadMoreButton}
    </>
  );
}

export function NewCards() {
  const getNewCardsQuery = useGetNewCardsFirstPageQuery();
  const projectName = useProjectName();
  return getNewCardsQuery.isFetching || getNewCardsQuery.data == null ? (
    <Spinner />
  ) : (
    <div>
      {Object.keys(getNewCardsQuery.data).length > 0 ? (
        <>
          <p>
            Check out the new cards added to {projectName} in the last two
            weeks.
          </p>
          {Object.entries(getNewCardsQuery.data).map(
            ([sourceKey, firstPage], index) => (
              <>
                {index != 0 && <hr />}
                <NewCardsForSource
                  key={sourceKey}
                  sourceKey={sourceKey}
                  firstPage={firstPage}
                />
              </>
            )
          )}
        </>
      ) : (
        <p>
          Looks like nothing was added to {projectName} in the last two weeks.
          :(
        </p>
      )}
    </div>
  );
}
