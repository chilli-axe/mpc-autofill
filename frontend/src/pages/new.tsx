import Head from "next/head";
import React from "react";

import { ProjectName } from "@/common/constants";
import { NoBackendDefault } from "@/components/NoBackendDefault";
import { NewCards } from "@/features/new/New";
import Footer from "@/features/ui/Footer";
import { ProjectContainer } from "@/features/ui/Layout";
import {
  useBackendConfigured,
  useProjectName,
} from "@/store/slices/backendSlice";

function NewOrDefault() {
  const backendConfigured = useBackendConfigured();
  return backendConfigured ? (
    <>
      <h1>What&apos;s New?</h1>
      <NewCards />
      <Footer />
    </>
  ) : (
    <NoBackendDefault />
  );
}

export default function New() {
  const projectName = useProjectName();
  return (
    <ProjectContainer>
      <Head>
        <title>{`${projectName} New Cards`}</title>
        <meta
          name="description"
          content={`Check out the new cards added to ${ProjectName} over the last two weeks.`}
        />
      </Head>
      <NewOrDefault />
    </ProjectContainer>
  );
}
