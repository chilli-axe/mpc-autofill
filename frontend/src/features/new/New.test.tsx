import { screen, waitFor } from "@testing-library/react";

import {
  cardDocument1,
  cardDocument2,
  cardDocument3,
  cardDocument4,
  sourceDocument1,
} from "@/common/test-constants";
import { renderWithProviders } from "@/common/test-utils";
import { NewCards } from "@/features/new/New";
import {
  defaultHandlers,
  newCardsFirstPageNoResults,
  newCardsFirstPageWithTwoSources,
  newCardsPageForSource1,
} from "@/mocks/handlers";
import { server } from "@/mocks/server";

test("the html structure of the new cards page with two sources, each with result/s", async () => {
  server.use(newCardsFirstPageWithTwoSources, ...defaultHandlers);
  const rendered = renderWithProviders(<NewCards />);
  await waitFor(() =>
    expect(screen.getByText(sourceDocument1.name)).toBeInTheDocument()
  );
  expect(
    screen.getByText("Check out the new cards", { exact: false }).parentElement!
  ).toMatchSnapshot();
});

test("the html structure of the new cards page with no data", async () => {
  server.use(newCardsFirstPageNoResults, ...defaultHandlers);
  const rendered = renderWithProviders(<NewCards />);
  await waitFor(() =>
    expect(screen.getByText(":(", { exact: false })).toBeInTheDocument()
  );
  expect(
    screen.getByText(":(", { exact: false }).parentElement!
  ).toMatchSnapshot();
});

test("clicking to show another page of results in the new cards page", async () => {
  server.use(
    newCardsFirstPageWithTwoSources,
    newCardsPageForSource1,
    ...defaultHandlers
  );
  renderWithProviders(<NewCards />);
  await waitFor(() =>
    expect(screen.getByText(sourceDocument1.name)).toBeInTheDocument()
  );
  expect(screen.getByText(cardDocument1.name)).toBeInTheDocument();
  expect(screen.getByText(cardDocument2.name)).toBeInTheDocument();
  expect(screen.queryByText(cardDocument3.name)).not.toBeInTheDocument();
  expect(screen.queryByText(cardDocument4.name)).not.toBeInTheDocument();

  screen.getByText("Load More").click();

  await waitFor(() =>
    expect(screen.getByText(cardDocument3.name)).toBeInTheDocument()
  );
  expect(screen.getByText(cardDocument4.name)).toBeInTheDocument();
  expect(screen.queryByText("Load More")).not.toBeInTheDocument();
});
