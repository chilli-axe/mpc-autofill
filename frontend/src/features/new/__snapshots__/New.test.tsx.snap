// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of the new cards page with no data 1`] = `
<div>
  <p>
    Looks like nothing was added to 
    Test Site
     in the last two weeks. :(
  </p>
</div>
`;

exports[`the html structure of the new cards page with two sources, each with result/s 1`] = `
.c3 {
  z-index: 1;
  opacity: 1;
}

.c3:hover {
  cursor: pointer;
}

.c2 {
  z-index: 0;
  background: #4e5d6c;
  border: solid 0px black;
}

.c0 {
  display: inline;
}

.c1 {
  display: inline;
}

<div>
  <p>
    Check out the new cards added to 
    Test Site
     in the last two weeks.
  </p>
  <h3
    class="c0 orpheus"
  >
    <em>
      Source 1
    </em>
  </h3>
     
  <p
    class="c1 text-primary"
  >
    4
     new card
    s
  </p>
  <div
    class="g-0 row row-cols-xxl-6 row-cols-lg-4 row-cols-md-3 row-cols-sm-2 row-cols-2"
  >
    <div
      class="col"
    >
      <div
        class="mpccard mpccard-hover card"
      >
        <div
          class="pb-0 text-center card-header"
        >
          <p
            class="mpccard-slot"
          >
            1st January, 2000
          </p>
        </div>
        <div>
          <div
            class="c2 rounded-lg shadow-lg ratio ratio-7x5"
          >
            <img
              alt="Card 1"
              class="c3 card-img card-img-fade-in"
              data-nimg="fill"
              decoding="async"
              loading="lazy"
              src=""
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
            />
          </div>
          <div
            class="mb-0 text-center card-body"
          >
            <div
              class="mpccard-name card-subtitle h6"
            >
              Card 1
            </div>
            <div
              class="mpccard-spacing"
            >
              <p
                class="mpccard-source card-text"
              >
                Source 1 [1200 DPI]
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col"
    >
      <div
        class="mpccard mpccard-hover card"
      >
        <div
          class="pb-0 text-center card-header"
        >
          <p
            class="mpccard-slot"
          >
            1st January, 2000
          </p>
        </div>
        <div>
          <div
            class="c2 rounded-lg shadow-lg ratio ratio-7x5"
          >
            <img
              alt="Card 2"
              class="c3 card-img card-img-fade-in"
              data-nimg="fill"
              decoding="async"
              loading="lazy"
              src=""
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
            />
          </div>
          <div
            class="mb-0 text-center card-body"
          >
            <div
              class="mpccard-name card-subtitle h6"
            >
              Card 2
            </div>
            <div
              class="mpccard-spacing"
            >
              <p
                class="mpccard-source card-text"
              >
                Source 1 [1200 DPI]
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <br />
  <div
    class="d-grid gap-0 mx-auto"
    style="max-width: 20%;"
  >
    <button
      class="btn btn-primary"
      type="button"
    >
      Load More
    </button>
  </div>
  <hr />
  <h3
    class="c0 orpheus"
  >
    <em>
      Source 2
    </em>
  </h3>
     
  <p
    class="c1 text-primary"
  >
    1
     new card
  </p>
  <div
    class="g-0 row row-cols-xxl-6 row-cols-lg-4 row-cols-md-3 row-cols-sm-2 row-cols-2"
  >
    <div
      class="col"
    >
      <div
        class="mpccard mpccard-hover card"
      >
        <div
          class="pb-0 text-center card-header"
        >
          <p
            class="mpccard-slot"
          >
            1st January, 2000
          </p>
        </div>
        <div>
          <div
            class="c2 rounded-lg shadow-lg ratio ratio-7x5"
          >
            <img
              alt="Card 5"
              class="c3 card-img card-img-fade-in"
              data-nimg="fill"
              decoding="async"
              loading="lazy"
              src=""
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
            />
          </div>
          <div
            class="mb-0 text-center card-body"
          >
            <div
              class="mpccard-name card-subtitle h6"
            >
              Card 5
            </div>
            <div
              class="mpccard-spacing"
            >
              <p
                class="mpccard-source card-text"
              >
                Source 2 Cardbacks [1200 DPI]
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <br />
</div>
`;
